Add Array Support

Shelnutt2 opened the pull request:
This add Array Support, Array Schema, Dimension, Domain, Attribute, Group and adjust VFS to store context on struct.

Test coverage is only about 58% as query support is not implemented so any code paths that require data/non-empty arrays can't be tested via unit testing yet.

For functions which take void pointers to handle arrays of varies datatypes I've handled that by taking interface{}'s as a parameter to the function and then switching on the domain/dimension datatype to be able to convert the golang array to a void * in C.

Closes #7.

jakebolewski commented on dimension.go:
```diff
@@ -0,0 +1,381 @@
+	var cname *C.char = C.CString(name)
+	defer C.free(unsafe.Pointer(cname))
+
+	if reflect.TypeOf(domain).Kind() != reflect.Slice {
```
Is there a way here to check if the domain slice type == extent slice type?  Probably best to enforce this as issues with auto type conversion can be tricky to debug.

Shelnutt2 replied:
A safety check for this is simple to add, I'll do that next.

jakebolewski commented:
Few minor comments, looks great!

Shelnutt2 commented:
I have adjusted the LoadArraySchema to be a function that is not associated to the struct. There are no static functions in golang for being member of a struct. Having this create and read the function will solve other issues of having partial structs due to bad reads, and having the already initialized struct. 

I realized there were a few important functions missing from the ArraySchema class, I added those also.

The config's LoadFromFile function is also a struct member, I see in the c++ api the construct just takes an optional path for loading from file. Unless you object I'll make the same change to the `NewConfig()` function so it takes an optional path and loads config from disk that way.

jakebolewski commented:
👍 sounds good to me
